index.js

```diff
@@ -1,85 +1,125 @@
-        filesystemURI === this.state.imgURI) {
+      if (this.props.source.uri === this.state.imgURI || filesystemURI === this.state.imgURI) {
         return null;
       }
       await this.loadImage(filesystemURI, this.props.source.uri);
     }
   }
 
+  async componentWillUnmount() {
+    this._interaction && this._interaction.cancel();
+    this.mounted = false;
+    await this.checkClear();
+  }
+
+  async checkClear() {
+    try {
+      if (this.downloadResumable) {
+        console.log('Image download interrupted');
+        const t = await this.downloadResumable.pauseAsync();
+        const filesystemURI = await this.getImageFilesystemKey(this.props.source.uri);
+        const metadata = await FileSystem.getInfoAsync(filesystemURI);
+        if (metadata.exists) {
+          await FileSystem.deleteAsync(t.fileUri);
+        }
+      }
+    } catch (error) {
+      console.log(error);
+    }
+  }
+
   async getImageFilesystemKey(remoteURI) {
-    const hashed = await Crypto.digestStringAsync(
-      Crypto.CryptoDigestAlgorithm.SHA256,
-      remoteURI
-    );
-    return `${FileSystem.cacheDirectory}${hashed}`;
+    const hashed = await Crypto.digestStringAsync(Crypto.CryptoDigestAlgorithm.SHA256, remoteURI);
+    return `${FileSystem.documentDirectory}${hashed}`;
   }
 
   async loadImage(filesystemURI, remoteURI) {
+    if (this.downloadResumable && this.downloadResumable._removeSubscription) {
+      this.downloadResumable._removeSubscription();
+    }
     try {
       // Use the cached image if it exists
       const metadata = await FileSystem.getInfoAsync(filesystemURI);
       if (metadata.exists) {
         this.setState({
-          imgURI: filesystemURI
+          imgURI: filesystemURI,
         });
         return;
       }
 
       // otherwise download to cache
-      const imageObject = await FileSystem.downloadAsync(
+      this.downloadResumable = FileSystem.createDownloadResumable(
         remoteURI,
-        filesystemURI
+        filesystemURI,
+        {},
+        (dp) => this.onDownloadUpdate(dp)
       );
-      this.setState({
-        imgURI: imageObject.uri
-      });
+
+      const imageObject = await this.downloadResumable.downloadAsync();
+      if (this.mounted) {
+        if (imageObject && imageObject.status == '200') {
+          this.setState({
+            imgURI: imageObject.uri,
+          });
+        }
+      }
     } catch (err) {
-      console.log('Image loading error:', err);
-      this.setState({imgURI: remoteURI});
+      console.log('Image download error:', err);
+      if (this.mounted) {
+        this.setState({ imgURI: null });
+      }
+      const metadata = await FileSystem.getInfoAsync(filesystemURI);
+      if (metadata.exists) {
+        await FileSystem.deleteAsync(filesystemURI);
+      }
+    }
+  }
+
+  onDownloadUpdate(downloadProgress) {
+    if (downloadProgress.totalBytesWritten >= downloadProgress.totalBytesExpectedToWrite) {
+      console.log('Download completed');
```

Comment:
We shouldn't log success, don't want to clutter the console

Reply:
Alright, removed the unnecessary comments